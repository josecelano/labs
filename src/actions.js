/**
 * Add three actions here:
 *
 * - checkItem
 * - addItem
 * - editInput
 */

export function checkItem(text) {
    return {type: 'CHECK_ITEM', text: text}
}

export function addItem(text) {
    return {type: 'ADD_ITEM', text: text}
}

export function editInput(inputValue) {
    return {type: 'EDIT_INPUT', inputValue: inputValue}
}