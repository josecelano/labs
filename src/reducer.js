/**
 * Write your reducer in here that handles the three actions from actions.js and has two properties in its state:
 *
 * - items      [Array]  The items in the todolist
 * - inputValue [String] The value of the input field
 */

const initialState = {
    items: [
        {
            text: 'Get to Las Palmas 🌴',
            checked: true
        }, {
            text: 'Learn the basics of React 👩‍🎓',
            checked: false
        }, {
            text: 'Go surfing 🏄',
            checked: false
        }
    ],
    inputValue: ''
};

const appState = (state = initialState, action) => {
    switch (action.type) {
        case 'ADD_ITEM':
            return {
                ...state,
                items: [
                    ...state.items, {
                        text: action.text,
                        checked: false
                    }
                ],
                inputValue: ''
            }
        case 'EDIT_INPUT':
            return {
                ...state,
                inputValue: action.inputValue
            }
        case 'CHECK_ITEM':
            return {
                ...state,
                items: state
                    .items
                    .map(item => item.text === action.text
                        ? {
                            text: action.text,
                            checked: !item.checked
                        }
                        : item)
            }
        default:
            return state
    }
}

export default appState;