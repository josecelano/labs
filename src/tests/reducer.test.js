/**
 * Tests for the reducer
 */
import reducer from '../reducer';
import {checkItem, addItem, editInput} from '../actions';

describe('reducer', () => {
    it('should return the initial state', () => {
        expect(reducer(undefined, {})).toEqual({
            items: [
                {
                    text: 'Get to Las Palmas 🌴',
                    checked: true
                }, {
                    text: 'Learn the basics of React 👩‍🎓',
                    checked: false
                }, {
                    text: 'Go surfing 🏄',
                    checked: false
                }
            ],
            inputValue: ''
        })
    })

    it('should handle the checkItem action', () => {
        expect(reducer(undefined, checkItem('Get to Las Palmas 🌴'))).toEqual({
            items: [
                {
                    text: 'Get to Las Palmas 🌴',
                    checked: false
                }, {
                    text: 'Learn the basics of React 👩‍🎓',
                    checked: false
                }, {
                    text: 'Go surfing 🏄',
                    checked: false
                }
            ],
            inputValue: ''
        })
    })

    it('should handle the addItem action', () => {
        expect(reducer(undefined, addItem())).toEqual({
            items: [
                {
                    text: 'Get to Las Palmas 🌴',
                    checked: true
                }, {
                    text: 'Learn the basics of React 👩‍🎓',
                    checked: false
                }, {
                    text: 'Go surfing 🏄',
                    checked: false
                }, {
                    text: '',
                    checked: false
                }
            ],
            inputValue: ''
        })
    })

    it('should handle the editInput action', () => {
        expect(reducer(undefined, editInput('input text'))).toEqual({
            items: [
                {
                    text: 'Get to Las Palmas 🌴',
                    checked: true
                }, {
                    text: 'Learn the basics of React 👩‍🎓',
                    checked: false
                }, {
                    text: 'Go surfing 🏄',
                    checked: false
                }
            ],
            inputValue: 'input text'
        })
    })
})