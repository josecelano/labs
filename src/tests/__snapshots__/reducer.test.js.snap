exports[`test should add an item based on the inputValue in the Redux state when the addItem action is dispatched 1`] = `
Object {
  "inputValue": "",
  "items": Array [
    Object {
      "checked": true,
      "text": "Get to Las Palmas 🌴",
    },
    Object {
      "checked": false,
      "text": "Learn the basics of React 👩‍🎓",
    },
    Object {
      "checked": false,
      "text": "Go surfing 🏄",
    },
    Object {
      "checked": false,
      "text": undefined,
    },
  ],
}
`;

exports[`test should check an item based on its text when the checkItem action is dispatched 1`] = `
Object {
  "inputValue": "",
  "items": Array [
    Object {
      "checked": true,
      "text": "Get to Las Palmas 🌴",
    },
    Object {
      "checked": true,
      "text": "Learn the basics of React 👩‍🎓",
    },
    Object {
      "checked": false,
      "text": "Go surfing 🏄",
    },
  ],
}
`;

exports[`test should edit the inputValue in the state when the editInput action is dispatched 1`] = `
Object {
  "inputValue": undefined,
  "items": Array [
    Object {
      "checked": true,
      "text": "Get to Las Palmas 🌴",
    },
    Object {
      "checked": false,
      "text": "Learn the basics of React 👩‍🎓",
    },
    Object {
      "checked": false,
      "text": "Go surfing 🏄",
    },
  ],
}
`;

exports[`test should return the initial state when called without a state 1`] = `
Object {
  "inputValue": "",
  "items": Array [
    Object {
      "checked": true,
      "text": "Get to Las Palmas 🌴",
    },
    Object {
      "checked": false,
      "text": "Learn the basics of React 👩‍🎓",
    },
    Object {
      "checked": false,
      "text": "Go surfing 🏄",
    },
  ],
}
`;

exports[`test should return the state when no action is passed 1`] = `
Object {
  "inputValue": "Something",
  "items": Array [],
}
`;
